stage1/init/common/mount.go

```diff
@@ -238,3 +242,150 @@ func ensureDestinationExists(source, destination string) error {
+ * playground in "/tmp/rkt.propagate.XXXX" and create the bind mount in
+ * "/tmp/rkt.propagate.XXXX/mount" with the correct read-only attribute before
+ * moving it.
+ *
+ * Another complication is that the playground cannot be on a shared mount
+ * because Linux does not allow MS_MOVE to be applied to mounts with MS_SHARED
+ * parent mounts. But by default, systemd mounts everything as shared, see:
+ * https://github.com/systemd/systemd/blob/v231/src/core/mount-setup.c#L392
+ * We set up the temporary playground as a slave bind mount to avoid this
+ * limitation.
+ */
+func AppAddOneMount(p *stage1commontypes.Pod, ra *schema.RuntimeApp, sourcePath string, dstPath string, readOnly bool, enterCmd []string) {
+	/* Prepare a temporary playground that is not a shared mount */
+	playgroundMount, err := ioutil.TempDir("", "rkt.propagate.")
+	if err != nil {
+		log.FatalE("error creating temporary propagation directory", err)
+		os.Exit(1)
+	}
+	defer os.Remove(playgroundMount)
+
+	err = syscall.Mount(playgroundMount, playgroundMount, "bind", syscall.MS_BIND, "")
+	if err != nil {
+		log.FatalE("error mounting temporary directory", err)
+		os.Exit(1)
+	}
+	defer syscall.Unmount(playgroundMount, 0)
+
+	err = syscall.Mount("", playgroundMount, "none", syscall.MS_SLAVE, "")
+	if err != nil {
+		log.FatalE("error mounting temporary directory", err)
+		os.Exit(1)
+	}
+
+	/* Bind mount the source into the playground, possibly read-only */
+	mountTmp := filepath.Join(playgroundMount, "mount")
+	if err := os.MkdirAll(mountTmp, 0700); err != nil {
+		log.FatalE("error creating temporary mount directory", err)
+		os.Exit(1)
+	}
+	defer os.Remove(mountTmp)
+
+	err = syscall.Mount(sourcePath, mountTmp, "bind", syscall.MS_BIND, "")
+	if err != nil {
+		log.FatalE("error mounting temporary directory", err)
+		os.Exit(1)
+	}
+	defer syscall.Unmount(mountTmp, 0)
+
+	if readOnly {
+		err = syscall.Mount("", mountTmp, "bind", syscall.MS_REMOUNT|syscall.MS_RDONLY|syscall.MS_BIND, "")
+		if err != nil {
+			log.FatalE("error remounting temporary mount directory read-only", err)
+			os.Exit(1)
+		}
+	}
+
+	/* Now that the bind mount has the correct attributes (RO or RW), move
+	 * it to the propagation directory prepared by systemd-nspawn */
+	mountOutside := filepath.Join("/run/systemd/nspawn/propagate/", "rkt-"+p.UUID.String(), "rkt.mount")
+	mountInside := filepath.Join("/run/systemd/nspawn/incoming/", filepath.Base(mountOutside))
+	/* TODO(Alban): instead of always using mkdir(), we should check if the
+	 * source is a file or a directory, and use mkdir() or touch() accordingly. */
+	if err := os.MkdirAll(mountOutside, 0700); err != nil {
+		log.FatalE("error creating temporary mount directory", err)
+		os.Exit(1)
+	}
+	defer os.Remove(mountOutside)
+
+	err = syscall.Mount(mountTmp, mountOutside, "", syscall.MS_MOVE, "")
+	if err != nil {
+		log.FatalE("error moving mount directory", err)
+		os.Exit(1)
+	}
+	defer syscall.Unmount(mountOutside, 0)
```

Comment:
The "defer" instructions should be executed in reverse order.

It seems to work now. I must have executed an old version. Sorry for the noise.